File: 프로그래머스/DFS,BFS 네트워크.cpp
#include <string>
#include <vector>

using namespace std;

bool visited[201] = { false };

vector<int> v[201];

void makeVisited(int pos) {
	visited[pos] = true;
	int s = v[pos].size();
	for (int i = 0; i<s; i++) {
		if (visited[v[pos][i]] == false) {
			makeVisited(v[pos][i]);
		}
	}
}

int checkGroup(int pos) {
	if (visited[pos] == true) { //이미 이전 노드들과 연결되었었다면
		return 0;
	}
	else { //처음이라면 연결된 노드들도 true만들어야함
		makeVisited(pos);
		return 1;
	}
}

int solution(int n, vector<vector<int>> computers) {
	int answer = 0;

	for (int i = 0; i<n; i++) {
		for (int k = 0; k<n; k++) {
			if (computers[i][k] == 1) {
				v[i].push_back(k);
			}
		}
	}

	for (int i = 0; i<n; i++) {
		answer += checkGroup(i);
	}


	return answer;
}

File: 프로그래머스/DFS,BFS 네트워크.java
package algorithm;


import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

class Solution {
	int parent[];
	int getParent(int x) {
		if(parent[x]==x) {
			return x;
		}
		else {
			return parent[x] = getParent(parent[x]);
		}
	}
	void unionParent(int a, int b) {
		int A,B;
		A = getParent(a);
		B = getParent(b);
		if(A!=B) {
            //System.out.println(A+" "+B);
			if(A>B) 
                parent[B] = A;
			else
				parent[A] = B;
		}
	}
	
	
    public int solution(int n, int[][] computers) {
        int answer = 0;
        parent = new int[n+1];
        for(int i=0;i<n;i++) {
        	parent[i] = i;
        }
        
        Map<Integer,Integer> m = new HashMap<>();
     
        
        for(int i=0;i<n;i++) {
        	for(int j=0;j<n;j++) {
        		if(i!=j) {
        			if(computers[i][j]==1)
        				unionParent(i,j);
        		}
        	}
        }
        
        for(int i=0;i<n;i++) {
        	int temp = getParent(i);
        	if(!m.containsKey(temp)) {
        		m.put(temp,1);
        		//System.out.println(temp);
        	}
        }
        
        answer = m.size();
        
        return answer;
    }
}